File: 50- Longest increasing subsequence/JavaScript/solution3.js
// By using dynamic programming + binary search:

// Time complexity: O(nlogn)
// Space complexity: O(n)

function ceilIndex(arr, num){
  let left = 0
  let right = arr.length-1;
  while(left < right-1){
    let mid = Math.floor((left+right)/2);
    if(num == arr[mid])
      return mid;
    else if(num < arr[mid])
      right = mid;
    else
      left = mid;
  }
  return right;
}

function lis(arr){
  if(arr.length < 2) return arr.length;
  let subsequence = [];
  subsequence.push(arr[0]);
  for(let i = 1; i < arr.length; i++){
    if(arr[i] <= subsequence[0])
      subsequence[0] = arr[i];
    else if(arr[i] > subsequence[subsequence.length-1])
      subsequence.push(arr[i]);
    else
      subsequence[ceilIndex(subsequence, arr[i])] = arr[i];
  }
  return subsequence.length;
}

File: 50- Longest increasing subsequence/Java/solution3.java
// By using dynamic programming + binary search:

// Time complexity: O(nlogn)
// Space complexity: O(n)

static int ceilIndex(ArrayList<Integer> arr, int num){
  int left = 0;
  int right = arr.size()-1;
  while(left < right-1){
    int mid = (left+right)/2;
    if(num == arr.get(mid))
      return mid;
    else if(num < arr.get(mid))
      right = mid;
    else
      left = mid;
  }
  return right;
}

static int lis(int[] arr){
  if(arr.length < 2) return arr.length;
  ArrayList<Integer> subsequence = new ArrayList<>();
  subsequence.add(arr[0]);
  for(int i = 1; i < arr.length; i++){
    if(arr[i] <= subsequence.get(0))
      subsequence.set(0, arr[i]);
    else if(arr[i] > subsequence.get(subsequence.size()-1))
      subsequence.add(arr[i]);
    else
      subsequence.set(ceilIndex(subsequence, arr[i]), arr[i]);
  }
  return subsequence.size();
}


